build.rs

```diff
@@ -4,8 +4,108 @@
+    let body = format!(
+        "crate::run_wast(r#\"{}\"#, \"{}\")",
+        p.display(),
+        out.path.get(0).unwrap()
+    );
+
+    Some(Test {
+        name: testname.to_string(),
+        body: body.to_string(),
+    })
+}
+
+fn main() -> anyhow::Result<()> {
+    println!("cargo:rerun-if-changed=build.rs");
+    println!("cargo:rerun-if-changed=test/ignores.txt");
 
-fn main() {
     generate_wasi_tests::build();
     generate_emscripten_tests::build();
+
+    let out_dir = PathBuf::from(
+        env::var_os("OUT_DIR").expect("The OUT_DIR environment variable must be set"),
+    );
+    let ignores = build_ignores_from_textfile("tests/ignores.txt".into())?;
+    let mut out = Testsuite {
+        buffer: String::new(),
+        path: vec![],
+        ignores: ignores,
+    };
+
+    for compiler in &["singlepass", "cranelift", "llvm"] {
+        writeln!(out.buffer, "#[cfg(feature=\"backend-{}\")]", compiler);
+        writeln!(out.buffer, "#[cfg(test)]")?;
+        writeln!(out.buffer, "#[allow(non_snake_case)]")?;
+        with_test_module(&mut out, compiler, |mut out| {
+            with_test_module(&mut out, "spec", |out| {
+                let spec_tests = test_directory(out, "tests/spectests", wast_processor)?;
+                // Skip running spec_testsuite tests if the submodule isn't checked
+                // out.
+                // if spec_tests > 0 {
+                //     test_directory_module(
+                //         out,
+                //         "tests/spec_testsuite/proposals/simd",
+                //         wast_processor,
+                //     )?;
+                //     test_directory_module(
+                //         out,
+                //         "tests/spec_testsuite/proposals/multi-value",
+                //         wast_processor,
+                //     )?;
+                //     test_directory_module(
+                //         out,
+                //         "tests/spec_testsuite/proposals/reference-types",
+                //         wast_processor,
+                //     )?;
+                //     test_directory_module(
+                //         out,
+                //         "tests/spec_testsuite/proposals/bulk-memory-operations",
+                //         wast_processor,
+                //     )?;
+                // } else {
+                //     println!(
+                //         "cargo:warning=The spec testsuite is disabled. To enable, run `git submodule \
+                //     update --remote`."
+                //     );
+                // }
+                Ok(())
+            })?;
+            Ok(())
+        })?;
+    }
+
+    // println!("{}", out.buffer);
+    // std::process::exit(1);
+    // Write out our auto-generated tests and opportunistically format them with
+    // `rustfmt` if it's installed.
+    let output = out_dir.join("generated_tests.rs");
+    fs::write(&output, out.buffer)?;
+    drop(Command::new("rustfmt").arg(&output).status());
```

Comment:
Because `drop` almost never does anything notable, if we're intentionally using it for some reason, we should explain why we need it!

Reply:
Done!